### PR TITLE
start api schema

### LINKED ISSUE #86: Validate JSON payloads by JSON Schema
[JSON Schema](http://json-schema.org/) is currently a draft specification for defining validation of JSON documents by description in JSON itself.

https://spacetelescope.github.io/understanding-json-schema/index.html

### PR DESCRIPTION
**Pull request by AndyOGo**
fixes #86

### CODE REVIEW

**Comment by AndyOGo** [comment]
@mjenny 
Separating schema into separate files:
https://stackoverflow.com/questions/38925803/how-do-i-break-a-json-schema-into-separate-files

File: schemas/validators/length.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      "descpription": "Specifies the maximum length of the value of the component."
+    },
+    "msg": {
+      "type": "string",
+      "title": "Error Message",
+      "description": "The error message will be displayed if the value of the component is outside the minimum and/or maximum length."
+    }
+  },
+
+  "anyOf": [
```
Comment (AndyOGo) [inline schemas/validators/length.json]: @mjenny 
I didn't know that `anyOf` for `required` works this way

**Comment by AndyOGo** [comment]
@mjenny 
@lina-salih 
A lot is prepared, though the following stuff is still missing:

- [ ] add good [meta information of `title` and `description` properties](https://spacetelescope.github.io/understanding-json-schema/reference/generic.html#metadata)
- [ ] add reasonable [`required`](https://spacetelescope.github.io/understanding-json-schema/reference/object.html#required-properties) and [`additionalProperties`](https://spacetelescope.github.io/understanding-json-schema/reference/object.html#properties) checks
- [ ] allowed [combinations](https://spacetelescope.github.io/understanding-json-schema/reference/combining.html?highlight=ref#combining-schemas) have to be specified
- [ ] abstract reusable schemas
- [ ] check that all [references `$ref`](https://spacetelescope.github.io/understanding-json-schema/structuring.html#reuse) are correct
- [ ] enable automatic documentation generation
- [ ] enable validation of static all test screens in `src/test` through a npm script

File: schemas/components/Error.json [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    "children": {
+      "type": "array",
+      "title": "Children",
+      "description": "An array of child components.",
+
+      "items": [{
+        "$ref": "../types/children.json#"
+      }]
+    },
+    "contact": {
```
Comment (AndyOGo) [inline schemas/components/Error.json]: @mjenny 
@lina-salih 
Ich hätte gedacht ein array von `contacts` wäre besser.
Oder haben wir fix immer nur eine telefonnummer und email?

**Comment by AndyOGo** [comment]
Okay, most of our JSON Schema files are valid, except our `$ref`s don't seem to work properly.
https://github.com/jessedc/ajv-cli/issues/50

**Comment by AndyOGo** [comment]
mhm running `npm run lint:testdata` fails with:

```sh
> @axa-ch/aletheia@0.0.2-beta.18 lint:testdata /Users/axawinterthur/dev/aletheia
> babel-node build/scripts/validate-schema.js

Validate /Users/axawinterthur/dev/aletheia/src/test/_config.js
Validate /Users/axawinterthur/dev/aletheia/src/test/buttons.js
Errors found in /Users/axawinterthur/dev/aletheia/src/test/buttons.js
Error: schema is invalid: data should be object
    at Ajv.validateSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:183:16)
    at Ajv._addSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:312:10)
    at Ajv.validate (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:94:26)
    at tests.forEach (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:71:9)
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:67:7)
    at Module._compile (module.js:660:30)
    at loader (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:573:32)
Validate /Users/axawinterthur/dev/aletheia/src/test/dropdowns.js
Errors found in /Users/axawinterthur/dev/aletheia/src/test/dropdowns.js
Error: schema is invalid: data should be object
    at Ajv.validateSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:183:16)
    at Ajv._addSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:312:10)
    at Ajv.validate (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:94:26)
    at tests.forEach (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:71:9)
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:67:7)
    at Module._compile (module.js:660:30)
    at loader (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:573:32)
Validate /Users/axawinterthur/dev/aletheia/src/test/error.js
Errors found in /Users/axawinterthur/dev/aletheia/src/test/error.js
Error: schema is invalid: data should be object
    at Ajv.validateSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:183:16)
    at Ajv._addSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:312:10)
    at Ajv.validate (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:94:26)
    at tests.forEach (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:71:9)
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:67:7)
    at Module._compile (module.js:660:30)
    at loader (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:573:32)
Validate /Users/axawinterthur/dev/aletheia/src/test/expand-button.js
Validate /Users/axawinterthur/dev/aletheia/src/test/icon.js
Errors found in /Users/axawinterthur/dev/aletheia/src/test/icon.js
Error: schema is invalid: data should be object
    at Ajv.validateSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:183:16)
    at Ajv._addSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:312:10)
    at Ajv.validate (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:94:26)
    at tests.forEach (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:71:9)
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:67:7)
    at Module._compile (module.js:660:30)
    at loader (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:573:32)
Validate /Users/axawinterthur/dev/aletheia/src/test/inputs.js
Errors found in /Users/axawinterthur/dev/aletheia/src/test/inputs.js
Error: schema is invalid: data should be object
    at Ajv.validateSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:183:16)
    at Ajv._addSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:312:10)
    at Ajv.validate (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:94:26)
    at tests.forEach (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:71:9)
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:67:7)
    at Module._compile (module.js:660:30)
    at loader (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:573:32)
Validate /Users/axawinterthur/dev/aletheia/src/test/list-item.js
Errors found in /Users/axawinterthur/dev/aletheia/src/test/list-item.js
Error: schema is invalid: data should be object
    at Ajv.validateSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:183:16)
    at Ajv._addSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:312:10)
    at Ajv.validate (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:94:26)
    at tests.forEach (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:71:9)
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:67:7)
    at Module._compile (module.js:660:30)
    at loader (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:573:32)
Validate /Users/axawinterthur/dev/aletheia/src/test/markdown.js
Errors found in /Users/axawinterthur/dev/aletheia/src/test/markdown.js
Error: schema is invalid: data should be object
    at Ajv.validateSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:183:16)
    at Ajv._addSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:312:10)
    at Ajv.validate (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:94:26)
    at tests.forEach (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:71:9)
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:67:7)
    at Module._compile (module.js:660:30)
    at loader (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:573:32)
Validate /Users/axawinterthur/dev/aletheia/src/test/modaldialog.js
Errors found in /Users/axawinterthur/dev/aletheia/src/test/modaldialog.js
Error: schema is invalid: data should be object
    at Ajv.validateSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:183:16)
    at Ajv._addSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:312:10)
    at Ajv.validate (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:94:26)
    at tests.forEach (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:71:9)
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:67:7)
    at Module._compile (module.js:660:30)
    at loader (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:573:32)
Validate /Users/axawinterthur/dev/aletheia/src/test/multiple-select-list.js
Errors found in /Users/axawinterthur/dev/aletheia/src/test/multiple-select-list.js
Error: schema is invalid: data should be object
    at Ajv.validateSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:183:16)
    at Ajv._addSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:312:10)
    at Ajv.validate (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:94:26)
    at tests.forEach (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:71:9)
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:67:7)
    at Module._compile (module.js:660:30)
    at loader (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:573:32)
Validate /Users/axawinterthur/dev/aletheia/src/test/radios.js
Errors found in /Users/axawinterthur/dev/aletheia/src/test/radios.js
Error: schema is invalid: data should be object
    at Ajv.validateSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:183:16)
    at Ajv._addSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:312:10)
    at Ajv.validate (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:94:26)
    at tests.forEach (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:71:9)
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:67:7)
    at Module._compile (module.js:660:30)
    at loader (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:573:32)
Validate /Users/axawinterthur/dev/aletheia/src/test/screens/fehlerScreen.js
Validate /Users/axawinterthur/dev/aletheia/src/test/screens/mfAuswahlFahrzeug.js
Validate /Users/axawinterthur/dev/aletheia/src/test/screens/mfFahrzeuglenker.js
Validate /Users/axawinterthur/dev/aletheia/src/test/screens/mfOrtZeit.js
Validate /Users/axawinterthur/dev/aletheia/src/test/screens/mfSchadenhergang.js
Validate /Users/axawinterthur/dev/aletheia/src/test/screens/mfSchadenObjekte.js
Validate /Users/axawinterthur/dev/aletheia/src/test/screens/mfSchuld.js
Validate /Users/axawinterthur/dev/aletheia/src/test/select-list.js
Errors found in /Users/axawinterthur/dev/aletheia/src/test/select-list.js
Error: schema is invalid: data should be object
    at Ajv.validateSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:183:16)
    at Ajv._addSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:312:10)
    at Ajv.validate (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:94:26)
    at tests.forEach (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:71:9)
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:67:7)
    at Module._compile (module.js:660:30)
    at loader (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:573:32)
Validate /Users/axawinterthur/dev/aletheia/src/test/test-list.js
Validate /Users/axawinterthur/dev/aletheia/src/test/test.js
Validate /Users/axawinterthur/dev/aletheia/src/test/typos.js
Errors found in /Users/axawinterthur/dev/aletheia/src/test/typos.js
Error: schema is invalid: data should be object
    at Ajv.validateSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:183:16)
    at Ajv._addSchema (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:312:10)
    at Ajv.validate (/Users/axawinterthur/dev/aletheia/node_modules/ajv/lib/ajv.js:94:26)
    at tests.forEach (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:71:9)
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:67:7)
    at Module._compile (module.js:660:30)
    at loader (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:144:5)
    at Object.require.extensions.(anonymous function) (/Users/axawinterthur/.nvm/versions/node/v9.4.0/lib/node_modules/babel-cli/node_modules/babel-register/lib/node.js:154:7)
    at Module.load (module.js:573:32)
Validate /Users/axawinterthur/dev/aletheia/src/test/vorstellungReview.js

/Users/axawinterthur/dev/aletheia/build/scripts/validate-schema.js:84
  throw errors;
  ^
[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @axa-ch/aletheia@0.0.2-beta.18 lint:testdata: `babel-node build/scripts/validate-schema.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the @axa-ch/aletheia@0.0.2-beta.18 lint:testdata script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/axawinterthur/.npm/_logs/2018-05-22T12_41_12_058Z-debug.log
```

**Comment by AndyOGo** [comment]
@mjenny 
@lina-salih 
@LucaMele 
Hmh, seems that reliable tools for automatic documentation generation of JSON Schema files in markdown format aren't available at NPM (only via git clone and manual install...)
HTML format seems to be available, but then we need to serve it ourselves. If we use markdown, we can serve it through github 🤔

**Comment by mjenny** [comment]
@AndyOGo
I would go for the markdown approach since IMHO using github seems to be the proper way to do it.

**Comment by AndyOGo** [comment]
Could improve enums
https://github.com/json-schema-org/json-schema-spec/issues/57

**Comment by mjenny** [comment]
Looks promising in regards to the readability of enums.

**Comment by AndyOGo** [comment]
@lina-salih 
@mjenny 
@LucaMele 
Hmh, seems the `@adobe/jsonschema2md` has some bugs:

- [`default` not supported](https://github.com/adobe/jsonschema2md/issues/68)
- [`enum`s don't work randomly](https://github.com/adobe/jsonschema2md/issues/69)

**Comment by LucaMele** [comment]
please rebase son i can test with the current `npm run build-lib`

**Comment by AndyOGo** [comment]
@LucaMele 
@mjenny 
@lina-salih 
I have provided 2 PRs to `@adobe/jsonschema2md` for [**enum**'s in nested properties](https://github.com/adobe/jsonschema2md/pull/70) and [missing **default** values](https://github.com/adobe/jsonschema2md/pull/71).
Docs are regenerated:)
Now we should just double check all schema, make sure they are correct and merge this magic 🎉 

PS:
Please upvote my PR's

**Comment by LucaMele** [comment]
Wow this is a BIG PR! What is exactly the goal here? To have a interface and predefined schema with iWARP?

**Comment by AndyOGo** [comment]
Yeah, it is quite some work.

The goal is:

- enabling automatic validation of JSON payloads emitted and received by the UI-renderer 🎈 
- automatic validation of our test files 🍰 
- automatic documentation generation in markdown 🎉 
- IMHO help prevent finger-pointing from iWarp at UI-Renderer 😠

**Comment by AndyOGo** [comment]
@LucaMele 
@lina-salih 
@mjenny 
Unfortunately I just found other bugs:

- [refs in items of type array do not work](https://github.com/adobe/jsonschema2md/issues/72)
- [`allOf`, `anyOf`, `oneOf` etc. do not work](https://github.com/adobe/jsonschema2md/issues/52)

Actually here are [all issues](https://github.com/adobe/jsonschema2md/issues)...

And Adobe requires a **signed CLA** for accepting contributions...
https://github.com/adobe/jsonschema2md/blob/master/Contributing.md#contributor-license-agreement

**Pull request merged**